executor/src/witgen/machines/block_machine.rs

```diff
@@ -309,60 +331,57 @@ impl<T: FieldElement> BlockMachine<T> {
+                        for (poly_id, next, constraint) in inner_value
                             .constraints
                             .into_iter()
-                            .map(|(poly, constraint)| {
-                                (poly.poly_id(), poly.name.clone(), poly.next, constraint)
-                            })
+                            .map(|(poly, constraint)| (poly.poly_id(), poly.next, constraint))
                             .collect::<Vec<_>>()
                         {
-                            self.handle_constraint(
-                                poly_id as usize,
-                                name,
-                                next,
-                                constraint,
-                                &mut change_logger,
-                            );
+                            progress_in_last_step |=
+                                self.handle_constraint(poly_id as usize, next, constraint);
+                        }
+
+                        change_logger.log_constraints(self, &outer_constraints, false);
+                        for constraint in &outer_constraints {
+                            if outer_assignments.constraints.contains(constraint) {
+                                println!("Outer constraint: {:?}", constraint);
+                                panic!("Duplicate (outer)!");
+                            }
+                        }
+                        self.handle_outer_constraints(&outer_constraints, &mut left_mut);
+
+                        progress_in_last_step |= !outer_constraints.is_empty();
+                        outer_assignments.constraints.extend(outer_constraints);
+
+                        if progress_in_last_step {
+                            progress_steps.push(step);
                         }
                     }
                 }
                 Err(e) => errors.push(format!("In row {}: {e}", self.row).into()),
             }
+            processing_sequence_iterator.report_progress(progress_in_last_step);
         }
+
+        log::trace!("End processing block machine");
+
         // Only succeed if we can assign everything.
         // Otherwise it is messy because we have to find the correct block again.
-        let unknown_variables = left
+        let success = left_mut
             .iter()
-            .filter_map(|l| l.as_ref().ok().map(|l| l.nonzero_variables()))
-            .concat()
-            .iter()
-            .cloned()
-            .collect::<BTreeSet<_>>();
-        let value_assignments = outer_assignments
-            .constraints
-            .iter()
-            .filter_map(|(var, con)| match con {
-                Constraint::Assignment(_) => Some(*var),
-                Constraint::RangeConstraint(_) => None,
-            })
-            .collect::<BTreeSet<_>>();
+            .all(|v| v.as_ref().ok().map(|ae| ae.is_constant()) == Some(true));
```

Comment:
@chriseth Note that this is different from the previous behavior, where incomplete elements were ignored. But I think in the success case, there shouldn't be any incomplete clauses on the lefthand side, right?